### PR TITLE
Authority overload monitor

### PR DESCRIPTION
**Pull request by halfprice**
## Description 

This PR implements overload monitor in AuthorityState.

Overload monitor periodically monitors 3 signals:
  - the execution queueing latency
  - txn enqueue rate
  - txn execution rate

to make a decision whether it should enter load shedding mode, and what percentage of traffic to shed.

Currently, we store the load shedding information in AuthorityState::overload_info struct, and expose it using
metrics. This PR doesn't implement any actions when an authority is in load shedding mode.

## Test Plan 

1. Added unit tests to test overload_monitor module

2. Single node benchmark `cargo run --release --bin sui-single-node-benchmark -- --component with-tx-manager no-move`

with this PR: `Execution finished in 24.127s, TPS=20723.670576532517`
before this PR: `Execution finished in 24.007s, TPS=20827.258716207773`

3. local cluster testing to show metrics
Screenshot 2024-01-29 at 5 55 23 PM

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-core/src/consensus_throughput_calculator.rs [hunk 1]
```diff
@@ -469,6 +469,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg_attr(msim, ignore)]
```
Comment (halfprice) [inline crates/sui-core/src/consensus_throughput_calculator.rs:472]: Any test that manually creates an AuthorityMetrics fails `simtest` because of the code panics at this line where we initiate the start time: https://github.com/MystenLabs/sui/blob/ea5666b9c19141f34bb797d4b52c85b4e6a46441/crates/sui-core/src/metrics.rs#L105

So I turn all simtest off for those tests, and also it looks like they are all unit test.

Please let me know if there is a better way to fix this.

File: crates/sui-swarm-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap [hunk 2]
```diff
@@ -112,6 +112,17 @@ validator_configs:
       max_txn_age_in_queue:
         secs: 1
         nanos: 0
+      overload_monitor_interval:
```
Comment (halfprice) [inline crates/sui-swarm-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap:115]: I don't think this change is a breaking change for the operator because of the default used here: https://github.com/MystenLabs/sui/blob/ea5666b9c19141f34bb797d4b52c85b4e6a46441/crates/sui-config/src/node.rs#L167

File: crates/sui-core/src/consensus_throughput_calculator.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/consensus_throughput_calculator.rs:472]: Yes, unfortunately there is no way for simtest to rewrite or intercept regular `#[test]` tests, and if they do anything like checking the time, they will panic, because the simulator runtime has not been initialized.

You could just convert it to a `#[tokio::test]`, but adding the ignore is fine as well?

Comment (halfprice) [reply crates/sui-core/src/consensus_throughput_calculator.rs:472]: I'll just leave it as it then :)

File: crates/sui-core/src/authority.rs [hunk 3]
```diff
@@ -269,6 +274,12 @@ pub struct AuthorityMetrics {
     // Tracks recent average txn queueing delay between when it is ready for execution
     // until it starts executing.
     pub execution_queueing_latency: LatencyObserver,
+
+    // Tracks the rate of transactions become ready for execution in transaction manager.
+    pub txn_ready_rate_tracker: Arc<Mutex<RateTracker>>,
```
Comment (mystenmark) [inline crates/sui-core/src/authority.rs:279]: My guess is that contention on this mutex won't be an issue right now, because it should be locked very very briefly relative to overall execution time. But I am a bit nervous about adding something that could be an issue later, and which will probably not be easy to notice

File: crates/sui-swarm-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap (hunk 2 above)
Comment (mystenmark) [reply crates/sui-swarm-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap:115]: yes, this should be okay

File: crates/sui-core/src/overload_monitor.rs [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    // Deflate the execution rate to account for the case that execution_rate is close to
+    // txn_ready_rate.
+    if execution_rate * 0.9 > txn_ready_rate {
+        return 0;
+    }
+
+    // In order to maintain execution queue length, we need to drop at least (1 - executionRate / readyRate).
+    // TO reduce the queue length, here we add 10% more transactions to drop.
```
Comment (mystenmark) [inline crates/sui-core/src/overload_monitor.rs:114]: I'm not sure I understand the match here, especially the `0.9` and `0.1` constants. So, suppose we are 2x overloaded, e.g. our execution rate is 5 and our ready rate is 10.

Then we get (1.0 - 5 * 0.9 / 10) + 0.1 which is 0.65, so we are dropping 30% more transactions than we need to in order to make the incoming and outgoing rates equal.

File: crates/sui-core/src/overload_monitor.rs [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    true
+}
+
+// Calculates the percentage of transactions to drop in order to reduce execution queue.
+// Returns the integer percentage between 0 and 100.
+fn calculate_load_shedding_percentage(txn_ready_rate: f64, execution_rate: f64) -> u32 {
+    // When transaction ready rate is practically 0, we aren't adding more load to the
+    // execution driver, so no shedding.
+    // TODO: consensus handler or transaction manager can also be overloaded.
+    if txn_ready_rate < 1e-10 {
```
Comment (mystenmark) [inline crates/sui-core/src/overload_monitor.rs:103]: This seems like an oddly low threshold. I would not be surprised to see a lot of numerical instability at rates well above 1e-10, (say, 1e-5). So it's not obvious to me that we wouldn't see the system thrashing in and out of overload protection at very low traffic rates.

It would seem that a much higher threshold, say 10 or even 100 tps should be safe?

File: crates/sui-core/src/authority.rs (hunk 3 above)
Comment (halfprice) [reply crates/sui-core/src/authority.rs:284]: Currently it is not an issue because there is *almost* no concurrency to it.

Execution driver uses single thread to schedule and spawn new tasks, that's where the execution_rate_tracker is updated, and transaction manager also is entirely single threaded at the moment (every txn requires taking the TransactionManager::Inner lock for scheduling). The reason we need a Mutex here is for reading the rate in the overload_monitor, there we can control it to use relatively long update interval (10s currently) so that the contention for read and writing is rare.

I thought about what we are going to do in the future when execution driver and transaction manager can be multi-threaded. What we can do is to have a rate tracker in each thread to tracking the per-thread rate, and in the overload_monitor, we aggregate them, and maintain a low lock conflict.

I updated the comments to reflect this.

I also tried to only use atomic type to implement RateTracker, and the complexity there is non-trivial.

Please let me know what do you think?

File: crates/sui-core/src/overload_monitor.rs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    let load_shedding_percentage = min(
+        load_shedding_percentage,
+        config.max_load_shedding_percentage,
+    );
+    (overload_status, load_shedding_percentage)
+}
+
+#[cfg(test)]
```
Comment (mystenmark) [inline crates/sui-core/src/overload_monitor.rs:150]: I think we need a simulation test of some sort here. I don't mean an e2e simulator test, just a test with:

- An event emitter (aka "load generator") that generates events at a rate that varies over time
- A load shedder that sheds load according to the logic in this module
- An event consumer (aka "executor") that consumes events at some rate.

Then we can get some idea of what happens in the following scenarios
- load spike
- execution slow down
- rapid load oscillations / spikes
- very low traffic rates

The main thing we would want to see is whether the system produces oscillations on its own, and whether it can return to normal quickly after an adverse event is over.

But, this is just a suggestion, if you have a quicker / easier way to evaluate this let me know. I just think it will be very time consuming to do these tests in a test cluster with a load generator. (We'll need to do that as well to fully validate this, but that will have a much slower feedback loop and we may not be able to test as many scenarios)

. Otherwise I don't know how to begin reasoning about what kind of behavior this system might lead to. For instance

Basically set up a scenario with a randomized load generator and

File: crates/sui-core/src/overload_monitor.rs [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    // In order to maintain execution queue length, we need to drop at least (1 - executionRate / readyRate).
+    // TO reduce the queue length, here we add 10% more transactions to drop.
+    (((1.0 - execution_rate * 0.9 / txn_ready_rate) + 0.1).min(1.0) * 100.0).round() as u32
+}
+
+// Given overload signals (`queueing_latency`, `txn_ready_rate`, `execution_rate`), return whether
+// the authority server should enter load shedding mode, and how much percentage of transactions to drop.
+fn check_overload_signals(
```
Comment (mystenmark) [inline crates/sui-core/src/overload_monitor.rs:120]: One concern I have is, suppose for some reason we have a queueing delay that is unrelated to load. I can't imagine what would cause this, but perhaps we add such a bug in the future. It seems that in this case, the system would never recover no matter how much load we shed, so we would simply enter a death spiral. Is there anything we can do about that?

File: crates/sui-core/src/overload_monitor.rs (hunk 4 above)
Comment (halfprice) [reply crates/sui-core/src/overload_monitor.rs:114]: Yeah, I'm also not 100% confident with these const. Your calculation is mostly right.

0.9 is account for the case where execution rate ~= ready rate. In a steady state, these two rates are probably going to oscillate. One moment, it's (1997, 2001), and another moment, it's (2003, 1998) . So here I want to make sure that when we say ready rate is higher, is indeed higher. Ideally, the 0.9 should be replaced with oscillation rate in a steady state.

For the 0.1, it's for the case where we want the queueing delay to drop. In your example, simply dropping 50% of requests only make enqueue and dequeue rate to match. So we need a bit more to make it to drop. I also thought about maybe use *1.1 instead of +0.1. I don't know which one is better tbo.

I think these two are those parameters that it's hard to find the best value. So I use conservative numbers to start with. My plan is to set it here and tweak it when I test it in private testnet (hopefully one or two rounds of test will show whether they are ok or not).

File: crates/sui-core/src/overload_monitor.rs (hunk 6 above)
Comment (halfprice) [reply crates/sui-core/src/overload_monitor.rs:150]: 100% agree!

My plan was to write such test after I finish the actual shedding of the transaction. But I think I can create one that checks system status. Later on when shedding is implemented, I can evolve that test to further see the effectiveness of shedding.

Let me do it in a separate PR.

**Review by halfprice — COMMENTED** [review]
Thank you so much @mystenmark for your comment!

File: crates/sui-core/src/overload_monitor.rs (hunk 7 above)
Comment (halfprice) [reply crates/sui-core/src/overload_monitor.rs:120]: Yes, the `max_load_shedding_percentage` in the configuration is our big red button. When such thing happens, we can turn off this feature by setting `max_load_shedding_percentage` to 0.

File: crates/sui-core/src/authority.rs (hunk 3 above)
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:284]: I think what I missed before is that it is only locked for reading in the overload monitor task, which runs infrequently, so I agree this is fine

File: crates/sui-core/src/overload_monitor.rs (hunk 5 above)
Comment (mystenmark) [reply crates/sui-core/src/overload_monitor.rs:103]: ok, makes sense for now

File: crates/sui-core/src/overload_monitor.rs (hunk 6 above)
Comment (mystenmark) [reply crates/sui-core/src/overload_monitor.rs:150]: sounds good

**Review by mystenmark — APPROVED** [review]
